### ISSUE TITLE
[MNT] Add joblib backend option and set default to all parallelized estimators

### ISSUE THREAD

**Issue by baraline** [aeon-toolkit/aeon#1797]
### Describe the issue

Some of the estimators that use joblib for parallelization use process-based backend, while other use threads-based backend. Ideally, we want this to be a parameter tunable by the users. To this end, we want to include a `joblib_backend` parameter which would default to `threading`, and use this parameter to set the joblib `backend` parameter during `Parallel` calls.

**Comment by TonyBagnall** [comment]
I remember we looked at this in some detail when debugging some rocket features, was there some interaction with its use of prange? I cant remember!

```python

def _static_transform_uni(X, parameters, indices):
    """Transform a 2D collection of univariate time series.

    Implemented separately to the multivariate version for numba efficiency reasons.
    See issue #1778.
    """
    n_cases, n_timepoints = X.shape
    (
        _,
        _,
        dilations,
        n_features_per_dilation,
        biases,
    ) = parameters
    n_kernels = len(indices)
    n_dilations = len(dilations)
    f = n_kernels * np.sum(n_features_per_dilation)
    features = np.zeros((n_cases, f), dtype=np.float32)
    for i in prange(n_cases):
```

**Comment by CodeLionX** [comment]
I think adding this parameter to so many APIs would clutter the parameter lists. What are your thoughts on using [`joblib.parallel_config`](https://joblib.readthedocs.io/en/latest/generated/joblib.parallel_config.html#joblib.parallel_config)?
This means that all aeon code does not specify any parallel backend in their calls to `joblib.Parallel` (if it does not require a certain backend) and the user can choose its backend by wrapping calls within:

```python
with joblib.parallel_config(backend="loky"):
   # aeon code
```

The same is actually possible for `n_jobs` etc.

**Comment by baraline** [comment]
True, forgot about this option ! That or we set a global variable such as `AEON_JOBLIB_BACKEND` and document its usage

**Comment by CodeLionX** [comment]
A custom env var `AEON_JOBLIB_BACKEND` IMO just makes sense if it is also used for other parallel stuff besides joblib. Otherwise, we can use the existing (documented and potentially already known) joblib-facilities.

**Comment by MatthewMiddlehurst** [comment]
Sounds like a good idea, would have to be documented though. Not sure about removing `n_jobs`.

**Comment by baraline** [comment]
I'm OK with the use of the `parallel_config` option and let user warp estimators if they want to change from the default. But if we want to default to threads, we need to do the following, which look like  `parallel_config` doesn't overwrite :

```
with parallel_config(backend='loky'):
    p = Parallel(backend='threading')
    print(p._backend)
```
You obtain `<joblib._parallel_backends.ThreadingBackend object at 0x000001D91BB731F0>`

This means that if we want the `parallel_config` option to work, we would need to use default `Parallel()`, which use the loky backend. This is exactly the one we want to move away from for multiple reasons (check https://scikit-learn.fondation-inria.fr/joblib-sprint/ for some). 

Or did I miss something ?

**Comment by baraline** [comment]
So for options we have :

- Add a parameter to specify backend, which defaults to threads for all estimators with parallel capabilities
- Add an environment variable to specify backend package wide.

With an equal level of documentation, I think the first options would allow us to be more flexible if for whatever reason we need to use processes for some estimators and threads for others. Also it should be less prone to causing issues with existing methods.

But true, this adds a bit of parameter bloat to the API... We could introduce a dictionary regrouping all joblib params and use it as kwargs when calling parallel ?

Other ideas are welcome !

**Comment by CodeLionX** [comment]
I'm also in favor of 1. How many joblib parameters do we have? If there are just 2, I would not use kwargs for that.

**Comment by SrijanShovit** [comment]
Hey @baraline Can I take up this issue?

**Comment by baraline** [comment]
Hi, you can, however I would first require an analysis of the current state of the use of parallelism across modules, as some PR have worked on some module (distance, classification) we would first need a clear view of the remaining stuff to do to have an uniform interface for parallelism

**Comment by SrijanShovit** [comment]
sure I will go through the repo and provide the same

**Comment by SrijanShovit** [comment]
@baraline Could you let me know if understanding is correct?

To confirm once more, there are some files with class based code and constructor has the param: `parallel_backend=None` and this is being used in Parallel (from joblib). This is what we need as consistent across entire repo.

<img width="1218" height="1154" alt="Image" src="https://github.com/user-attachments/assets/8aa33ec1-d916-4b11-89e9-9bf25229a5e0" />

<img width="1048" height="296" alt="Image" src="https://github.com/user-attachments/assets/44119782-d16b-41d6-acfa-f83a37931ae1" />

Below is a screenshot of a file that is going the hardcoded way:

<img width="1096" height="1142" alt="Image" src="https://github.com/user-attachments/assets/27aa50a8-3ca1-4d4c-940e-ffb6bbaa10bb" />

<img width="1150" height="730" alt="Image" src="https://github.com/user-attachments/assets/ae98c0e3-d243-4ae9-9035-85f44b4d8e35" />

Once confirmed, I will proceed to create an organized list of modules consistent and inconsistent with use of parallelism provided by user.

**Comment by baraline** [comment]
That would be the first step yes. But I'm wondering if this uniformlisation is necessary after all... 

Consider the case were we expose a backend parameter to choose between thread and processes. 
It's likely that one class might have been designed with threads in mind rather than processes. 

Allowing to change that behavior by a parameter seem a bit weird, as the best choice should be made by the class developer.
 @aeon-toolkit/aeon-infrastructure-workgroup any thought on that matter? The initial though of the issue is a bit lost in time and I'm not sure it is relevant now...

**Comment by SrijanShovit** [comment]
@baraline For the time being, let me know if there is some other issue that needs help currently. Anyways, I have gone through the whole repo in the last hour.

**Comment by SrijanShovit** [comment]
Hey @baraline Following up on this

> That would be the first step yes. But I'm wondering if this uniformlisation is necessary after all...
> 
> Consider the case were we expose a backend parameter to choose between thread and processes. It's likely that one class might have been designed with threads in mind rather than processes.
> 
> Allowing to change that behavior by a parameter seem a bit weird, as the best choice should be made by the class developer. @aeon-toolkit/aeon-infrastructure-workgroup any thought on that matter? The initial though of the issue is a bit lost in time and I'm not sure it is relevant now...

**Comment by SebastianSchmidl** [comment]
> Anyways, I have gone through the whole repo in the last hour.

@SrijanShovit What were the results of your analysis? At least we can document the scope of this issue if you already went through the hassle of checking the codebase.

**Comment by SrijanShovit** [comment]
@SebastianSchmidl These are the files where already `parallel_backend` is passed in constructor:

- aeon/transformations/collection/feature_based/_catch22.py
- aeon/transformations/collection/interval_based/_random_intervals.py
- aeon/transformations/collection/interval_based/_supervised_intervals.py
- aeon/transformations/collection/shapelet_based/_shapelet_transform.py
- aeon/classification/dictionary_based/_redcomets.py
- aeon/classification/distance_based/_proximity_forest.py
- aeon/classification/feature_based/_tdmvdc.py
- aeon/base/_estimators/interval_based/base_interval_forest.py

Below are the ones with hardcoded backend:

### Thread Based
- aeon/regression/sklearn/_rotation_forest_regressor.py
- aeon/transformations/collection/dictionary_based/_sfa.py
- aeon/classification/convolution_based/_arsenal.py
- aeon/classification/dictionary_based/_boss.py
- aeon/classification/dictionary_based/_muse.py
- aeon/classification/dictionary_based/_tde.py
- aeon/classification/dictionary_based/_weasel_v2.py
- aeon/classification/dictionary_based/_weasel.py
- aeon/classification/early_classification/_teaser.py
- aeon/classification/early_classification/_probability_threshold.py
- aeon/classification/ordinal_classification/_ordinal_tde.py
- aeon/classification/sklearn/_rotation_forest_classifier.py

### Process Based
- aeon/distances/_distance.py
- aeon/clustering/averaging/_shift_scale_invariant_averaging.py
- aeon/classification/ordinal_classification/_ordinal_tde.py

Let me know if further help is required

**Comment by SebastianSchmidl** [comment]
Thank you! This is valuable information.
